src/test/java/org/jenkinsci/plugins/gitclient/JGitClientTest.java

```diff
@@ -0,0 +1,162 @@
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.rules.TemporaryFolder;
+
+import org.jvnet.hudson.test.Issue;
+
+/**
+ * JGit Client-specific tests. In their own Test file to ward off bloat in 
+ * GitClient and minimized parameterized and other test logic conditional 
+ * on Git implementation.
+ *
+ * @author Brian Ray
+ */
+public class JGitClientTest {
+    /* These tests are only for the JGit client. */
+    private static final String GIT_IMPL_NAME = "jgit";
+
+    /* Instance under test. */
+    private GitClient gitClient;
+
+    /* Lower level "porcelain" API when gitClient can't do something. */
+    private org.eclipse.jgit.api.Git gitAPI;
+
+    /* Repo garbage collector. */
+    private GC gc;
+
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    @Rule
+    public TemporaryFolder tempFolder = new TemporaryFolder();
+    private File repoRoot;
+
+    @Before
+    public void setGitClientEtc() throws IOException, InterruptedException {
+        repoRoot = tempFolder.newFolder();
+        gitClient = Git.with(TaskListener.NULL, new EnvVars())
+                       .in(repoRoot)
+                       .using(GIT_IMPL_NAME)
+                       .getClient();
+        FileRepository repo = (FileRepository) gitClient.getRepository();
+        // See comment up in imports as to why this is fully qualified.
+        gitAPI = org.eclipse.jgit.api.Git.wrap(repo);
+        gc = new GC(repo);
+
+        File gitDir = gitClient.withRepository((r, channel) -> r.getDirectory());
+        gitClient.init_()
+                 .workspace(repoRoot.getAbsolutePath())
+                 .execute();
+        assertTrue("Missing " + gitDir, gitDir.isDirectory());
+    }
+
+    private ObjectId commitFile(final String path, final String content, final String commitMessage) throws Exception {
+        createFile(path, content);
+        gitClient.add(path);
+        gitClient.commit(commitMessage);
+
+        List<ObjectId> headList = gitClient.revList(HEAD);
+        assertThat(headList.size(), is(greaterThan(0)));
+        return headList.get(0);
+    }
+
+    private void createFile(final String path, final String content) throws Exception {
+        File aFile = new File(repoRoot, path);
+        File parentDir = aFile.getParentFile();
+        if (parentDir != null) {
+            parentDir.mkdirs();
+        }
+        try (PrintWriter writer = new PrintWriter(aFile, "UTF-8")) {
+            writer.printf(content);
+        } catch (FileNotFoundException | UnsupportedEncodingException ex) {
+            throw new GitException(ex);
+        }
+    }
+
+    private void packRefs() throws IOException {
+        gc.packRefs();
+    }
+
+    // No flavor of GitClient has a tag(String) API, only tag(String,String). 
+    // But sometimes we want a lightweight a.k.a. non-annotated tag.
+    private void tag(String name) throws GitException {
```

Comment:
To make it clear in callers, could this be called:
```suggestion
    private void lightweightTag(String name) throws GitException {
```